File: app/schemas/message.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, computed_field


class ConversationCreate(BaseModel):
    participant_ids: list[int]
    title: str | None = None
    conversation_type: str = "direct"


class ConversationParticipantUserOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    full_name: str
    email: str
    role: str


class ConversationParticipantOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    conversation_id: int | None = None
    joined_at: datetime
    last_read_at: datetime | None = None
    user: ConversationParticipantUserOut | None = None


class MessageCreate(BaseModel):
    body: str
    reply_to_message_id: int | None = None


class MessageReplyOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    sender_id: int | None
    sender_name: str | None = None
    body: str
    created_at: datetime


class MessageDeliveryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    message_id: int
    user_id: int
    delivered_at: datetime | None = None
    read_at: datetime | None = None


class MessageAttachmentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    message_id: int
    uploaded_by_id: int | None = None
    filename: str
    original_filename: str
    mime_type: str
    file_size: int
    storage_path: str
    created_at: datetime

    @computed_field
    @property
    def is_image(self) -> bool:
        return self.mime_type.startswith("image/")

    @computed_field
    @property
    def download_url(self) -> str:
        return f"/api/v1/message-attachments/{self.id}/download"


class MessageOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    conversation_id: int
    sender_id: int | None
    sender_name: str | None = None

    reply_to_message_id: int | None = None
    reply_to: MessageReplyOut | None = None

    body: str

    created_at: datetime
    updated_at: datetime | None = None

    deliveries: list[MessageDeliveryOut] = Field(default_factory=list)

    attachments: list[MessageAttachmentOut] = Field(default_factory=list)


class ConversationLatestMessageOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    body: str
    sender_id: int | None = None
    sender_name: str | None = None
    created_at: datetime


class ConversationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    school_id: int | None
    title: str | None
    conversation_type: str
    created_by_id: int | None

    created_at: datetime
    updated_at: datetime | None = None

    participants: list[ConversationParticipantOut] = Field(default_factory=list)
    messages: list[MessageOut] = Field(default_factory=list)

    unread_count: int = 0
    latest_message: ConversationLatestMessageOut | None = None
    last_activity: datetime | None = None


class MarkConversationReadOut(BaseModel):
    success: bool
    conversation_id: int
    user_id: int
    last_read_at: datetime | None = None


class UnreadMessageCountOut(BaseModel):
    unread_count: int


class MessageAttachmentCreateOut(BaseModel):
    success: bool
    attachment: MessageAttachmentOut


class MessageAttachmentDownloadOut(BaseModel):
    id: int
    filename: str
    original_filename: str
    mime_type: str
    file_size: int
    storage_path: str

    @computed_field
    @property
    def is_image(self) -> bool:
        return self.mime_type.startswith("image/")

    @computed_field
    @property
    def download_url(self) -> str:
        return f"/api/v1/message-attachments/{self.id}/download"

File: mhike-school-web/types/message.ts
export type ConversationType =
    | "direct"
    | "group"
    | "class"
    | "broadcast";

export type SchoolMessageUser = {
    id: number;
    full_name: string;
    email: string;
    role: string;
};

export type MessageReply = {
    id: number;
    sender_id: number | null;
    body: string;
    created_at: string;
};

export type MessageDelivery = {
    id: number;
    message_id: number;
    user_id: number;
    delivered_at: string | null;
    read_at: string | null;
};

export type MessageAttachment = {
    id: number;
    message_id: number;
    uploaded_by_id?: number | null;
    filename: string;
    original_filename: string;
    mime_type: string;
    file_size: number;
    storage_path: string;
    created_at: string;
};

export type MessageAttachmentUploadResponse = {
    id: number;
    message_id: number;
    uploaded_by_id?: number | null;
    filename: string;
    original_filename: string;
    mime_type: string;
    file_size: number;
    storage_path: string;
    created_at?: string | null;
};

export type MessageAttachmentCreateResponse = {
    success: boolean;
    attachment: MessageAttachment;
};

export type MessageAttachmentDownload = {
    id: number;
    filename: string;
    original_filename: string;
    mime_type: string;
    file_size: number;
    storage_path: string;
};

export type Message = {
    id: number;
    conversation_id: number;
    sender_id: number | null;
    body: string;
    reply_to_message_id?: number | null;
    reply_to?: MessageReply | null;
    created_at: string;
    updated_at?: string | null;
    deliveries?: MessageDelivery[];
    attachments?: MessageAttachment[];
};

export type ConversationParticipantUser = {
    id: number;
    full_name: string;
    email: string;
    role: string;
};

export type ConversationParticipant = {
    id: number;
    user_id: number;
    conversation_id: number;
    joined_at?: string;
    last_read_at?: string | null;
    user?: ConversationParticipantUser | null;
};

export type ConversationLatestMessage = {
    id: number;
    body: string;
    sender_id?: number | null;
    created_at: string;
};

export type Conversation = {
    id: number;
    school_id: number | null;
    created_by_id: number | null;
    title: string | null;
    conversation_type: ConversationType;
    created_at: string;
    updated_at?: string | null;
    participants?: ConversationParticipant[];
    messages?: Message[];
    unread_count?: number;
    latest_message?: ConversationLatestMessage | null;
    last_activity?: string | null;
};

export type ConversationCreatePayload = {
    participant_ids: number[];
    title?: string | null;
    conversation_type: ConversationType;
};

export type MessageCreatePayload = {
    body: string;
    reply_to_message_id?: number | null;
};

export type MarkConversationReadResponse = {
    success: boolean;
    conversation_id: number;
    user_id: number;
    last_read_at: string | null;
};

export type UnreadMessageCountResponse = {
    unread_count: number;
};